### PR TITLE
Various changes and fixes to UNNEST.

### PR DESCRIPTION
**Pull request by gianm**
Native changes:

1) UnnestDataSource: Replace "column" and "outputName" with "virtualColumn".
   This enables pushing expressions into the datasource. This in turn
   allows us to do the next thing...

2) UnnestStorageAdapter: Logically apply query-level filters and virtual
   columns after the unnest operation. (Physically, filters are pushed down,
   when possible.) This is beneficial because it allows filters and
   virtual columns to reference the unnested column, and because it is
   consistent with how the join datasource works.

3) Various documentation updates, including declaring "unnest" as an
   experimental feature for now.

SQL changes:

1) Rename DruidUnnestRel (& Rule) to DruidUnnestRel (& Rule). The rel
   is simplified: it only handles the UNNEST part of a correlated join.
   Constant UNNESTs are handled with regular inline rels.

2) Rework DruidCorrelateUnnestRule to focus on pulling Projects from
   the left side up above the Correlate. Additionally, fix issues related to
   tracking correlation variables.New test testUnnestTwice verifies
   that this works even when two UNNESTs are stacked on the same table.

3) Include ProjectCorrelateTransposeRule from Calcite to encourage
   pushing mappings down below the left-hand side of the Correlate.

4) Add a new CorrelateFilterLTransposeRule and CorrelateFilterRTransposeRule
   to handle pulling Filters up above the Correlate. New tests
   testUnnestWithFiltersOutside and testUnnestTwiceWithFilters verify
   this behavior.

5) Require a context feature flag for SQL UNNEST, since it's undocumented.
   As part of this, also cleaned up how we handle feature flags in SQL.
   They're now hooked into EngineFeatures, which is useful because not
   all engines support all features.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidCorrelateUnnestRule.java
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidCorrelateUnnestRule.java:68]: nit: this javadoc needs updated i suppose

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/CastOperatorConversion.java
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/CastOperatorConversion.java:79]: this isn't probably that cool because it loses the array element type. I think maybe #13890 has a better solution for this by switching cast to use `Calcites.getColumnTypeForRelDataType`

**Review by clintropolis — COMMENTED** [review]
this seems nice to me, and fixes the problem I was having using unnest with array columns in #13803

Screenshot 2023-03-07 at 5 36 51 AM

**Review by somu-imply — COMMENTED** [review]
I have a general Q. These 2 queries have the same plan:
```
SELECT d3 FROM numFoo, UNNEST(MV_TO_ARRAY(dim3)) as unnested (d3) where dim2 IN ('a','ab')
SELECT d3 FROM (select * from numFoo where dim2 IN ('a','ab')), UNNEST(MV_TO_ARRAY(dim3)) as unnested (d3)
```
But these 2 queries have different plans, the selector filter inside the table for the first query below plans as an unnest over a QueryDataSource while the second does not
```
SELECT d3 FROM (select * from "numFoo" where dim2='a'), UNNEST(MV_TO_ARRAY(dim3)) as unnested(d3) 
SELECT d3 FROM numFoo, UNNEST(MV_TO_ARRAY(dim3)) as unnested (d3) where dim2='a'
```

**Comment by somu-imply** [comment]
I think the reason is that there is an additional level of LogicalProject between Correlate and LogicalFilter in the query where `SELECT d3 FROM (select * from druid.numfoo where dim2='a'), UNNEST(MV_TO_ARRAY(dim3)) as unnested (d3)` it plans to 
```
136:LogicalProject(d3=[$17])
  134:LogicalCorrelate(subset=[rel#135:Subset#7.NONE.[]], correlation=[$cor0], joinType=[inner], requiredColumns=[{3}])
    125:LogicalProject(subset=[rel#126:Subset#2.NONE.[]], __time=[$0], dim1=[$1], dim2=[CAST('a':VARCHAR):VARCHAR], dim3=[$3], dim4=[$4], dim5=[$5], dim6=[$6], d1=[$7], d2=[$8], f1=[$9], f2=[$10], l1=[$11], l2=[$12], cnt=[$13], m1=[$14], m2=[$15], unique_dim1=[$16])
      123:LogicalFilter(subset=[rel#124:Subset#1.NONE.[]], condition=[=($2, 'a')])
        9:LogicalTableScan(subset=[rel#122:Subset#0.NONE.[]], table=[[druid, numfoo]])
    130:Uncollect(subset=[rel#131:Subset#5.NONE.[]])
      128:LogicalProject(subset=[rel#129:Subset#4.NONE.[]], EXPR$0=[MV_TO_ARRAY($cor0.dim3)])
        12:LogicalValues(subset=[rel#127:Subset#3.NONE.[0]], tuples=[[{ 0 }]])
```

which does not match the CorrelateFilterTranspose rule (L or R). We might need to address that but can be done with a later PR as well

**Review by somu-imply — APPROVED** [review]
Overall LGTM ! We can merge this and then I can fix up the boundary conditions

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidCorrelateUnnestRule.java
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidCorrelateUnnestRule.java:68]: Yep. Updated it.

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/CastOperatorConversion.java
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/CastOperatorConversion.java:79]: Yeah, yours is better. Merged it in. Thanks.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java
Comment (somu-imply) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:2875]: Replace with `new Object[]{"1", useDefault ? null: ImmutableList.of(""), ImmutableList.of("1"), "1", "xx"}`

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:302]: nice :+1:

**Review by clintropolis — APPROVED** [review]
lgtm, :+1: after CI

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidUnnestRel.java
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidUnnestRel.java:50]: Oh, I didn't realize RelNode implements Cloneable. This hasn't been an issue for any of the other DruidRels; none of them implement `clone`. Calcite's builtin rels don't either, except for MutableRel, which isn't involved here. I think it's fine to not implement this. I'll add an impl to `DruidRel` that throws an exception and write a comment remarking that it isn't needed.

**Comment by somu-imply** [comment]
Thanks, there is also an unused import in `DruidQueryRel` which would also come in the checkstyle fix.

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:299]: actually I think we want to always add the filter as a post filter even when we push down since the pushed down filter returns the whole row if any values match, so we want the value matcher to clean it up to exact matches

Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:299]: Let's merge this once CI passes, I'll fix them up. I understand the case where it returns the full row even for a match in MVDs. I'll also need to remove the allowSet and minor other changes. Will followup as soon as this gets merged

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidRel.java
Comment (somu-imply) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidRel.java:113]: IntelliJ inspections is flagging it as it does not throw a CloneNotSupportedException

**Comment by somu-imply** [comment]
Additionally there is a CI failure for this test `testGroupByFloatMinExpressionVsVirtualColumnWithExplicitStringVirtualColumnTypedInput`.

The problem seems to trigger from running with `-Ddruid.generic.useDefaultValueForNull=false`. The fix for this is to change the lines
```
"minVc",
NullHandling.replaceWithDefault() ? Float.POSITIVE_INFINITY : null
```
with
```
"minVc",
Float.POSITIVE_INFINITY
```

Although I am unsure what caused this to change

**Comment by gianm** [comment]
> Although I am unsure what caused this to change

Ah. That's because ExpressionVirtualColumn now becomes a pass-through in cases where the expression doesn't do anything. So the behavior of `min` and `minVc` are now the same. I updated the test case to reflect the new expected behavior.

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:299]: Here's the patch https://gist.github.com/gianm/86d55fb6e5ee58c935152761a9df4a4d. Feel free to roll this into what you're doing @somu-imply.

Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:299]: PR with the fix is at: https://github.com/apache/druid/pull/13919. If you want to incorporate this into your own work then feel free and I will close my PR.

**Pull request merged**
